### PR TITLE
what's new: redux foundation

### PR DESCRIPTION
**Pull request by japie1235813**
* Motivation for features / changes
More note-reading part to _redux/  and build up the selector-reducer-action foundation
 
* Technical description of changes
  * Modify a bunch of visibility to outside _redux/ for its usage in tensorboard app
  * Leave tests for reducer cause actions is not brought into the picture yet.

### CODE REVIEW

File: tensorboard/webapp/notification_center/index.ts [hunk 1]
```diff
@@ -14,3 +14,4 @@ limitations under the License.
 ==============================================================================*/
 
 export {NotificationCenterModule} from './_notification_center_module';
+// export * from './_redux/notification_center_selector';
```
Comment (stephanwlee) [inline tensorboard/webapp/notification_center/index.ts:17]: remove me.

File: tensorboard/webapp/notification_center/_views/notification_center_container.ts [hunk 2]
```diff
@@ -30,16 +32,20 @@ const iconMap = new Map([[CategoryEnum.WHATS_NEW, 'info_outline_24px']]);
...
-        };
-      });
-    })
-  );
+  notificationNotes$: Observable<ViewNotificationExt[]> = observableOf([]);
+
+  constructor(private readonly store: Store<State>) {}
+
+  ngOnInit() {
+    this.notificationNotes$ = this.store.select(getNotifications).pipe(
```
Comment (stephanwlee) [inline tensorboard/webapp/notification_center/_views/notification_center_container.ts:40]: you don't have to do this in ngOnInit. Move this to L35 or somewhere similar.

File: tensorboard/webapp/notification_center/_redux/notification_center_actions.ts [hunk 3]
```diff
@@ -12,3 +12,16 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+
+/** @typehack */ import * as _typeHackModels from '@ngrx/store/src/models';
+/** @typehack */ import * as _typeHackStore from '@ngrx/store';
+
+/**
+ * Fires when the bell icon is clicked.
+ */
+export const notificationBellClicked = createAction(
+  '[Notification] Notification Bell Clicked',
+  props<Notification>()
```
Comment (stephanwlee) [inline tensorboard/webapp/notification_center/_redux/notification_center_actions.ts:26]: bell is clicked so you don't need to pass the Notification as the prop.

File: tensorboard/webapp/notification_center/_redux/BUILD [hunk 4]
```diff
@@ -27,25 +30,45 @@ tf_ts_library(
...
+        ":types",
+    ],
+)
+
 tf_ts_library(
     name = "_redux_test",
     testonly = True,
     srcs = [
-        "notification_center_reducers_test.ts",
+        # "notification_center_reducers_test.ts",
```
Comment (stephanwlee) [inline tensorboard/webapp/notification_center/_redux/BUILD:64]: Remove OR add the test.

File: tensorboard/webapp/notification_center/_redux/notification_center_actions.ts (hunk 3 above)
Comment (stephanwlee) [reply tensorboard/webapp/notification_center/_redux/notification_center_actions.ts:26]: Please fix me.

File: tensorboard/webapp/notification_center/_views/notification_center_container.ts [hunk 5]
```diff
@@ -30,16 +32,18 @@ const iconMap = new Map([[CategoryEnum.WHATS_NEW, 'info_outline_24px']]);
...
-    map((notifications) => {
-      return notifications.map((notification) => {
-        return {
-          ...notification,
-          icon: iconMap.get(notification.category) ?? null,
-        };
-      });
-    })
-  );
+  notificationNotes$: Observable<ViewNotificationExt[]> = this.store
```
Comment (stephanwlee) [inline tensorboard/webapp/notification_center/_views/notification_center_container.ts:35]: add `readonly` please.

File: tensorboard/webapp/notification_center/_views/notification_center_test.ts [hunk 6]
```diff
@@ -33,7 +50,14 @@ describe('notification center', () => {
...
       ],
       declarations: [NotificationCenterContainer, NotificationCenterComponent],
+      providers: [
+        provideMockStore({
+          initialState: [],
+        }),
+      ],
     }).compileComponents();
+    store = TestBed.inject<Store<State>>(Store) as MockStore<State>;
+    store.overrideSelector(selectors.getNotifications, notificationNotes);
```
Comment (stephanwlee) [inline tensorboard/webapp/notification_center/_views/notification_center_test.ts:60]: remove this line here and move it to L64. 

By reading spec inside `it` in L64, you should be able to tell whether your assertion is correct. So, instead, there, 
```ts
// minimal data structure to test the module.
store.overrideSelector(selectors.getNotifications, [
 {
    category: CategoryEnum.WHATS_NEW,
    dateInMs: 100,
    title: '2.4 release',
    content: 'some testable content',
  },
]);
```

Comment (japie1235813) [reply tensorboard/webapp/notification_center/_views/notification_center_test.ts]: Thanks for the suggestion! Addressed.

**Review by stephanwlee — APPROVED** [review]

**Pull request merged**